File: src/api/subsonic.ts
import { fetch } from "@tauri-apps/plugin-http";
import { Artist, Album, Track, SubsonicConfig, SearchResults } from "../types/subsonic";

const APP_NAME = "hagtamp";
const API_VERSION = "1.16.1";

function md5(input: string): string {
  // Simple MD5 using SubtleCrypto is async; we use a synchronous implementation
  // via the built-in approach for Subsonic token auth
  let h0 = 0x67452301, h1 = 0xefcdab89, h2 = 0x98badcfe, h3 = 0x10325476;
  const msg = new TextEncoder().encode(input);
  const padded = new Uint8Array(Math.ceil((msg.length + 9) / 64) * 64);
  padded.set(msg);
  padded[msg.length] = 0x80;
  const bits = BigInt(msg.length * 8);
  const view = new DataView(padded.buffer);
  view.setUint32(padded.length - 8, Number(bits & 0xffffffffn), true);
  view.setUint32(padded.length - 4, Number(bits >> 32n), true);

  const S = [7,12,17,22,7,12,17,22,7,12,17,22,7,12,17,22,5,9,14,20,5,9,14,20,5,9,14,20,5,9,14,20,4,11,16,23,4,11,16,23,4,11,16,23,4,11,16,23,6,10,15,21,6,10,15,21,6,10,15,21,6,10,15,21];
  const K = Array.from({length: 64}, (_, i) => Math.floor(Math.abs(Math.sin(i + 1)) * 2 ** 32) >>> 0);

  for (let i = 0; i < padded.length; i += 64) {
    const M = Array.from({length: 16}, (_, j) => view.getUint32(i + j * 4, true));
    let [a, b, c, d] = [h0, h1, h2, h3];
    for (let j = 0; j < 64; j++) {
      let F: number, g: number;
      if (j < 16)      { F = (b & c) | (~b & d); g = j; }
      else if (j < 32) { F = (d & b) | (~d & c); g = (5 * j + 1) % 16; }
      else if (j < 48) { F = b ^ c ^ d;           g = (3 * j + 5) % 16; }
      else             { F = c ^ (b | ~d);         g = (7 * j) % 16; }
      F = (F + a + K[j] + M[g]) >>> 0;
      a = d; d = c; c = b;
      b = (b + ((F << S[j]) | (F >>> (32 - S[j])))) >>> 0;
    }
    h0 = (h0 + a) >>> 0; h1 = (h1 + b) >>> 0; h2 = (h2 + c) >>> 0; h3 = (h3 + d) >>> 0;
  }

  return [h0, h1, h2, h3]
    .map(n => n.toString(16).padStart(8, "0").match(/../g)!.reverse().join(""))
    .join("");
}

function randomSalt(len = 6): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  return Array.from({length: len}, () => chars[Math.floor(Math.random() * chars.length)]).join("");
}

class SubsonicClient {
  private config: SubsonicConfig | null = null;

  setConfig(config: SubsonicConfig) {
    this.config = config;
  }

  private authParams(): string {
    if (!this.config) throw new Error("Subsonic not configured");
    const salt = randomSalt();
    const token = md5(this.config.password + salt);
    return `u=${encodeURIComponent(this.config.username)}&t=${token}&s=${salt}&v=${API_VERSION}&c=${APP_NAME}&f=json`;
  }

  private url(method: string, extra = ""): string {
    if (!this.config) throw new Error("Subsonic not configured");
    const base = this.config.serverUrl.replace(/\/$/, "");
    return `${base}/rest/${method}?${this.authParams()}${extra ? "&" + extra : ""}`;
  }

  private async get<T>(method: string, extra = ""): Promise<T> {
    const res = await fetch(this.url(method, extra));
    const json = await res.json() as { "subsonic-response": { status: string; error?: { code: number; message: string }; [key: string]: unknown } };
    const root = json["subsonic-response"];
    if (root.status !== "ok") throw new Error(root.error?.message ?? "Subsonic error");
    return root as T;
  }

  async ping(): Promise<void> {
    await this.get("ping");
  }

  async getArtists(): Promise<Artist[]> {
    const data = await this.get<{ artists: { index: { artist: Artist | Artist[] }[] } }>("getArtists");
    return data.artists.index.flatMap(idx =>
      Array.isArray(idx.artist) ? idx.artist : [idx.artist]
    );
  }

  async getArtist(id: string): Promise<Album[]> {
    const data = await this.get<{ artist: { album: Album | Album[] } }>("getArtist", `id=${id}`);
    const albums = data.artist.album;
    return Array.isArray(albums) ? albums : [albums];
  }

  async getAlbum(id: string): Promise<Track[]> {
    const data = await this.get<{ album: { song: Track | Track[] } }>("getAlbum", `id=${id}`);
    const songs = data.album.song;
    return Array.isArray(songs) ? songs : [songs];
  }

  async search3(query: string): Promise<SearchResults> {
    const data = await this.get<{
      searchResult3: {
        artist?: Artist | Artist[];
        album?: Album | Album[];
        song?: Track | Track[];
      }
    }>("search3", `query=${encodeURIComponent(query)}&artistCount=20&albumCount=20&songCount=50`);
    const r = data.searchResult3;
    return {
      artists: r.artist ? (Array.isArray(r.artist) ? r.artist : [r.artist]) : [],
      albums: r.album ? (Array.isArray(r.album) ? r.album : [r.album]) : [],
      tracks: r.song ? (Array.isArray(r.song) ? r.song : [r.song]) : [],
    };
  }

  getStreamUrl(id: string): string {
    return this.url("stream", `id=${id}&format=mp3`);
  }

  getCoverArtUrl(id: string, size = 300): string {
    return this.url("getCoverArt", `id=${id}&size=${size}`);
  }
}

export const subsonic = new SubsonicClient();

File: src/player/subsonic.rs
use anyhow::{anyhow, Result};
use serde::Deserialize;

#[derive(Clone, Debug)]
pub struct SubsonicConfig {
    pub server_url: String,
    pub username: String,
    pub password: String,
}

#[derive(Clone, Debug)]
pub struct Artist {
    pub id: String,
    pub name: String,
    pub album_count: u32,
}

#[derive(Clone, Debug)]
pub struct Album {
    pub id: String,
    pub name: String,
    pub artist: String,
    pub artist_id: String,
    pub year: Option<u32>,
    pub cover_art: Option<String>,
    pub song_count: u32,
}

#[derive(Clone, Debug)]
pub struct Track {
    pub id: String,
    pub title: String,
    pub artist: String,
    pub album: String,
    pub track: Option<u32>,
    pub duration: Option<u32>,
    pub cover_art: Option<String>,
}

// ── Raw Subsonic JSON types ──────────────────────────────────────────────────

#[derive(Deserialize)]
struct SubsonicResponse {
    #[serde(rename = "subsonic-response")]
    inner: ResponseInner,
}

#[derive(Deserialize)]
struct ResponseInner {
    status: String,
    error: Option<SubsonicError>,
    artists: Option<ArtistsWrapper>,
    artist: Option<ArtistDetail>,
    album: Option<AlbumDetail>,
    #[serde(rename = "searchResult3")]
    search_result3: Option<SearchResult3>,
}

#[derive(Deserialize)]
struct SubsonicError {
    message: String,
}

#[derive(Deserialize)]
struct ArtistsWrapper {
    index: Vec<ArtistIndex>,
}

#[derive(Deserialize)]
struct ArtistIndex {
    artist: Option<serde_json::Value>,
}

#[derive(Deserialize)]
struct ArtistDetail {
    album: Option<serde_json::Value>,
}

#[derive(Deserialize)]
struct AlbumDetail {
    song: Option<serde_json::Value>,
}

#[derive(Deserialize)]
struct SearchResult3 {
    song: Option<serde_json::Value>,
    artist: Option<serde_json::Value>,
    album: Option<serde_json::Value>,
}

#[derive(Deserialize, Clone)]
struct RawArtist {
    id: String,
    name: String,
    #[serde(rename = "albumCount", default)]
    album_count: u32,
}

#[derive(Deserialize, Clone)]
struct RawAlbum {
    id: String,
    name: String,
    artist: String,
    #[serde(rename = "artistId")]
    artist_id: String,
    year: Option<u32>,
    #[serde(rename = "coverArt")]
    cover_art: Option<String>,
    #[serde(rename = "songCount", default)]
    song_count: u32,
}

#[derive(Deserialize, Clone)]
struct RawSong {
    id: String,
    title: String,
    artist: String,
    album: String,
    track: Option<u32>,
    duration: Option<u32>,
    #[serde(rename = "coverArt")]
    cover_art: Option<String>,
}

// ── Client ───────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct SubsonicClient {
    config: SubsonicConfig,
    client: reqwest::Client,
}

impl SubsonicClient {
    pub fn new(config: SubsonicConfig) -> Self {
        Self {
            config,
            client: reqwest::Client::new(),
        }
    }

    fn auth_params(&self) -> String {
        let salt: String = (0..6).map(|_| {
            let idx = rand_u8() % 36;
            if idx < 10 { (b'0' + idx) as char } else { (b'a' + idx - 10) as char }
        }).collect();
        let token = format!("{:x}", md5::compute(format!("{}{}", self.config.password, salt)));
        format!(
            "u={}&t={}&s={}&v=1.16.1&c=hagtamp&f=json",
            urlencod(&self.config.username), token, salt
        )
    }

    fn url(&self, method: &str, extra: &str) -> String {
        let base = self.config.server_url.trim_end_matches('/');
        if extra.is_empty() {
            format!("{}/rest/{}?{}", base, method, self.auth_params())
        } else {
            format!("{}/rest/{}?{}&{}", base, method, self.auth_params(), extra)
        }
    }

    async fn get(&self, method: &str, extra: &str) -> Result<ResponseInner> {
        let resp: SubsonicResponse = self.client
            .get(self.url(method, extra))
            .send().await?
            .json().await?;
        let inner = resp.inner;
        if inner.status != "ok" {
            let msg = inner.error.map(|e| e.message).unwrap_or_else(|| "unknown error".into());
            return Err(anyhow!("{}", msg));
        }
        Ok(inner)
    }

    pub async fn ping(&self) -> Result<()> {
        self.get("ping", "").await?;
        Ok(())
    }

    pub async fn get_artists(&self) -> Result<Vec<Artist>> {
        let r = self.get("getArtists", "").await?;
        let Some(artists) = r.artists else { return Ok(vec![]) };
        let mut result = vec![];
        for idx in artists.index {
            let Some(val) = idx.artist else { continue };
            let items: Vec<RawArtist> = if val.is_array() {
                serde_json::from_value(val)?
            } else {
                vec![serde_json::from_value(val)?]
            };
            for a in items {
                result.push(Artist { id: a.id, name: a.name, album_count: a.album_count });
            }
        }
        Ok(result)
    }

    pub async fn get_artist(&self, id: &str) -> Result<Vec<Album>> {
        let r = self.get("getArtist", &format!("id={}", id)).await?;
        let Some(artist) = r.artist else { return Ok(vec![]) };
        let Some(val) = artist.album else { return Ok(vec![]) };
        let items: Vec<RawAlbum> = if val.is_array() {
            serde_json::from_value(val)?
        } else {
            vec![serde_json::from_value(val)?]
        };
        Ok(items.into_iter().map(|a| Album {
            id: a.id, name: a.name, artist: a.artist, artist_id: a.artist_id,
            year: a.year, cover_art: a.cover_art, song_count: a.song_count,
        }).collect())
    }

    pub async fn get_album(&self, id: &str) -> Result<Vec<Track>> {
        let r = self.get("getAlbum", &format!("id={}", id)).await?;
        let Some(album) = r.album else { return Ok(vec![]) };
        let Some(val) = album.song else { return Ok(vec![]) };
        let items: Vec<RawSong> = if val.is_array() {
            serde_json::from_value(val)?
        } else {
            vec![serde_json::from_value(val)?]
        };
        Ok(items.into_iter().map(|s| Track {
            id: s.id, title: s.title, artist: s.artist, album: s.album,
            track: s.track, duration: s.duration, cover_art: s.cover_art,
        }).collect())
    }

    pub async fn search(&self, query: &str) -> Result<Vec<Track>> {
        let r = self.get("search3", &format!(
            "query={}&songCount=50&albumCount=0&artistCount=0", urlencod(query)
        )).await?;
        let Some(sr) = r.search_result3 else { return Ok(vec![]) };
        let Some(val) = sr.song else { return Ok(vec![]) };
        let items: Vec<RawSong> = if val.is_array() {
            serde_json::from_value(val)?
        } else {
            vec![serde_json::from_value(val)?]
        };
        Ok(items.into_iter().map(|s| Track {
            id: s.id, title: s.title, artist: s.artist, album: s.album,
            track: s.track, duration: s.duration, cover_art: s.cover_art,
        }).collect())
    }

    pub fn stream_url(&self, id: &str) -> String {
        self.url("stream", &format!("id={}&format=raw", id))
    }

    pub fn cover_art_url(&self, id: &str) -> String {
        self.url("getCoverArt", &format!("id={}&size=300", id))
    }
}

fn urlencod(s: &str) -> String {
    s.chars().flat_map(|c| match c {
        'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => vec![c],
        _ => format!("%{:02X}", c as u32).chars().collect(),
    }).collect()
}

fn rand_u8() -> u8 {
    use std::time::{SystemTime, UNIX_EPOCH};
    let t = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().subsec_nanos();
    (t & 0xff) as u8
}
